Write for .prj files

joshday opened the pull request:
My text editor removed spaces at the end of lines, artificially making the diff bigger.

joshday commented on src/writer.jl:
```diff
@@ -161,11 +161,25 @@ function write(path::AbstractString, obj; force=false)
+            wkt = convert(GeoFormatTypes.ESRIWellKnownText, crs)
+            Base.write(paths.prj, wkt.val)
+        catch
+            @warn "Could not write .prj file.  ArchGDAL may need to be loaded."
```
If I understand correctly, ArchGDAL owns/pirates the `Base.convert` methods for types in GeoFormatTypes.  Is this a correct/acceptable warning?

evetion commented on src/writer.jl:
```diff
@@ -167,9 +167,18 @@ function write(path::AbstractString, obj; force=false)
-        crs = GeoInterface.crs(obj)
-        crs isa GeoFormatTypes.ESRIWellKnownText && Base.write(paths.prj, crs.val)
+    crs = try
+        GI.crs(geoms)
```
I think on getting the crs from the `obj` might be more valid than on the vector of geometries. This is how I now do it in https://github.com/evetion/GeoDataFrames.jl/blob/master/src/utils.jl#L19

joshday replied:
Done!

rafaqz commented on src/writer.jl:
```diff
@@ -161,11 +161,24 @@ function write(path::AbstractString, obj; force=false)
+    end
+    if !isnothing(crs)
+        try
+            crs isa GeoFormatTypes.ESRIWellKnownText && Base.write(paths.prj, crs.val)
```
Here we can actually run `convert(ESRIWellKnownText, crs)` if it isnt `ESRIWellKnownText`.

It needs ArchGDAL to be loaded to actually do the conversion (withsome sanctioned type-pyracy) - but  we should make GeoFormatTypes.jl specifically tell you to run `using ArchGDAL` in the error.

evetion replied:
I disagree here, we shouldn't depend on ArchGDAL like that.

rafaqz replied:
Sorry I don't get it... there is no dependency.

This just means it won't error if ArchGDAL is loaded, but will error if it isn't.

joshday replied:
I agree with @rafaqz.  The sanctioned ArchGDAL pirated `convert` methods are hard to discover.  I changed the code to run the `convert` but provide a useful warning (not error) when the conversion fails.

rafaqz reviewed (commented):
Thanks, just need .dbf now...

joshday commented on src/writer.jl:
```diff
@@ -161,11 +161,26 @@ function write(path::AbstractString, obj; force=false)
+    catch
+        nothing
+    end
+    if !isnothing(crs)
+        try
+            Base.write(paths.prj, convert(GeoFormatTypes.ESRIWellKnownText, crs).val)
+        catch
+            @warn ".prj write failure: Could not convert CRS of type `$(typeof(crs))` to " *
+            "`GeoFormatTypes.ESRIWellKnownText`.  `using ArchGDAL` may load the necessary " *
+            "`Base.convert` method."
+        end
+    end
```
How's this now?

rafaqz replied:
Looks good!

evetion replied:
This will always yield the following warning, because `ESRIWellKnownText` actually instantiates to `ESRIWellKnownText{GeoFormatTypes.Unknown}`. It should be `GeoFormatTypes.CRS()` here. We should also change https://github.com/JuliaGeo/Shapefile.jl/blob/main/src/handle.jl#L57 to include `GeoFormatTypes.CRS()` as first argument.

```
┌ Warning: .prj write failure: Could not convert CRS of type `ESRIWellKnownText{GeoFormatTypes.Unknown}` to `GeoFormatTypes.ESRIWellKnownText`.  `using ArchGDAL` may load the necessary `Base.convert` method.
```

evetion replied:
Might be good to check that our tests include reading and writing the same shapefile completely, so we at least hit this codepath once.

joshday replied:
We can't test the roundtrip write-then-read until we can write .dbf, but I made the change to use CRS and added a test for warning before `using ArchGDAL` and no warning after.

joshday commented:
Change are now:

1. Changed type of `handle.crs` to `GeoFormatTypes.ESRIWellKnownText{GeoFormatTypes.CRS}` (breaking change).
2. `.prj` gets written if the `GeoInterface.crs(obj)` can be converted to `GeoFormatTypes.ESRIWellKnownText{GeoFormatTypes.CRS}` 
3. Added GeoInterface methods for `Handle`.
4. `write` now supports any `AbstractGeometryCollectionTrait`.
5. Added ArchGDAL as a test dependency.

rafaqz commented:
One problem may be that `MultiPatch` is itself a `GeometryCollectionTrait` geom. The last of the test cases (that I avoided) has a `MultiPatch` object, it might be good to see what happens if you try to write it.

The reason I didn't implement `write` for `MultiPatch` as a `GeometryCollection` it is the translation between GeoInterface and `MultiPatch` is pretty weird (or I just don't understand the object properly), and I'm not sure how to generalise it.

joshday commented:
>  it might be good to see what happens if you try to write it.

```Julia
julia> Shapefile.write("temp.shp", h.shapes[1])
ERROR: `getgeom` not implemented for `MultiPatch`: open a github issue at JuliaGeo/Shapefile.jl if you need this.
```

Seems like a good error message to me.  I think it's ready to merge if it has your blessings.

rafaqz commented:
Ah OK I misunderstood the `write` `AbstractGeometryCollection` part I get it now.
